XenAdmin/Wizards/RollingUpgradeWizard/PlanActions/RpuUploadAndApplySuppPackPlanAction.cs

```diff
@@ -0,0 +1,275 @@
+ * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING 
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE 
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF 
+ * SUCH DAMAGE.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using XenAdmin.Actions;
+using XenAdmin.Core;
+using XenAPI;
+using XenAdmin.Diagnostics.Checks;
+using XenAdmin.Diagnostics.Problems.HostProblem;
+using XenAdmin.Network;
+using XenAdmin.Wizards.PatchingWizard;
+using XenAdmin.Wizards.PatchingWizard.PlanActions;
+
+
+namespace XenAdmin.Wizards.RollingUpgradeWizard.PlanActions
+{
+    class RpuUploadAndApplySuppPackPlanAction : PlanActionWithSession
+    {
+        private Host host;
+        private List<Host> hosts;
+        private string suppPackPath;
+        private Dictionary<Host, Pool_update> uploadedSuppPacks;
+        public List<PlanAction> DelayedPlanActions;
+        private Dictionary<string, livepatch_status> livePatchStatus;
+
+        public RpuUploadAndApplySuppPackPlanAction(IXenConnection connection, Host host, List<Host> hosts, string path,
+            Dictionary<Host, Pool_update> uploadedUpdates)
+            : base(connection)
+        {
+            this.host = host;
+            this.hosts = hosts;
+            suppPackPath = path;
+            uploadedSuppPacks = uploadedUpdates;
+            DelayedPlanActions = new List<PlanAction>();
+            livePatchStatus = new Dictionary<string, livepatch_status>();
+        }
+
+        protected override void RunWithSession(ref Session session)
+        {
+            var conn = session.Connection;
+            var master = Helpers.GetMaster(conn);
+            var suppPackName = Path.GetFileName(suppPackPath);
+
+            // upload
+            UploadSuppPack(master, conn, session, suppPackName);
+
+            if (uploadedSuppPacks.ContainsKey(master))
+            {
+                var update = uploadedSuppPacks[master];
+                if (update != null)
+                {
+                    // precheck
+                    bool alreadyApplied;
+                    PrecheckSuppPack(update, suppPackName, out alreadyApplied);
+                    if (alreadyApplied)
+                    {
+                        RemoveSuppPackFromMaster(session, master, suppPackName, update);
+                        return;
+                    }
+
+                    // apply
+                    ApplySuppPack(conn, session, suppPackName, update);
+
+                    // after apply guidance
+                    AfterApplySuppPack(update);
+
+                    // remove from master
+                    RemoveSuppPackFromMaster(session, master, suppPackName, update);
+                }
+            }
+        }
+
+        private void UploadSuppPack(Host master, IXenConnection connection, Session session, string suppPack)
+        {
+            if (!uploadedSuppPacks.ContainsKey(master))

```

Comment:
If this plan action is retried because of a failure in the precheck phase, then the upload phase is correctly skipped, but the upload progress step also disappears from the plan actions progress view.